File: easy/301-400/339/rust/src/main.rs
// Copyright 2022 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use lazy_static::lazy_static;
use regex::Regex;

lazy_static! {
    static ref INFO_PATTERN: Regex =
        Regex::new(r"^(?P<name>[^:].{19})(?P<age>\d{2})(?P<birthday>\d{6})$").unwrap();
    static ref EXTENSION_PATTERN: Regex =
        Regex::new(r"::EXT::(?P<name>.{4})(?P<value>.{17})").unwrap();
}

#[cfg(not(tarpaulin_include))]
fn main() {
    println!("rad");
}

fn find_highest_salary(input: &str) -> String {
    todo!()
}

#[cfg(not(tarpaulin_include))]
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_highest_salary() {
        assert_eq!(
            "Marcelo Candela, $47706".to_string(),
            find_highest_salary(
                "Boyce Calles        83460319
::EXT::SAL 00000000000044722
Marcelo Candela     29040821
::EXT::JOB loser
::EXT::SAL 00000000000047706
Milton Camper       32541106"
            )
        );
    }
}

File: easy/301-400/339/go/main.go
// Copyright 2021 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var infoPattern = regexp.MustCompile(`^(?P<name>[^:].{19})(?P<age>\d{2})(?P<birthday>\d{6})$`)
var extensionPattern = regexp.MustCompile(`::EXT::(?P<name>.{4})(?P<value>.{17})`)

var zPrint = fmt.Println

func main() {
	_, _ = zPrint("hello world")
}

func findHighestSalary(input []string) (highest string) {
	highestSalary := -1
	highestSalaryOwner := ""
	currentName := ""
	for _, line := range input {
		matches := extensionPattern.FindStringSubmatch(line)
		if 0 == len(matches) {
			matches = infoPattern.FindStringSubmatch(line)
			currentName = strings.Trim(matches[1], " ")
		} else {
			extension := strings.Trim(matches[1], " ")
			if "SAL" == extension {
				salary, _ := strconv.Atoi(strings.Trim(matches[2], " "))
				if highestSalary < salary {
					highestSalary = salary
					highestSalaryOwner = currentName
				}
			}
		}
	}
	return fmt.Sprintf("%s, $%d", highestSalaryOwner, highestSalary)
}
